pkg/limayaml/validate.go

```diff
@@ -18,41 +18,63 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+func validateFile(f File, fieldName string) error {
+	if !strings.Contains(f.Location, "://") {
+		if _, err := localpathutil.Expand(f.Location); err != nil {
+			return fmt.Errorf("field `%s.location` refers to an invalid local file path: %q: %w", fieldName, f.Location, err)
+		}
+		// f.Location does NOT need to be accessible, so we do NOT check os.Stat(f.Location)
+	}
+	switch f.Arch {
+	case X8664, AARCH64, RISCV64:
+	default:
+		return fmt.Errorf("field `%s.arch` must be %q or %q, got %q", fieldName, X8664, AARCH64, f.Arch)
+	}
+	if f.Digest != "" {
+		if !f.Digest.Algorithm().Available() {
+			return fmt.Errorf("field `%s.digest` refers to an unavailable digest algorithm", fieldName)
+		}
+		if err := f.Digest.Validate(); err != nil {
+			return fmt.Errorf("field `%s.digest` is invalid: %s: %w", fieldName, f.Digest.String(), err)
+		}
+	}
+	return nil
+}
+
 func Validate(y LimaYAML, warn bool) error {
 	switch *y.Arch {
-	case X8664, AARCH64:
+	case X8664, AARCH64, RISCV64:
 	default:
-		return fmt.Errorf("field `arch` must be %q or %q , got %q", X8664, AARCH64, *y.Arch)
+		return fmt.Errorf("field `arch` must be %q, %q or %q, got %q", X8664, AARCH64, RISCV64, *y.Arch)
 	}
 
 	if len(y.Images) == 0 {
 		return errors.New("field `images` must be set")
 	}
 	for i, f := range y.Images {
-		if !strings.Contains(f.Location, "://") {
-			if _, err := localpathutil.Expand(f.Location); err != nil {
-				return fmt.Errorf("field `images[%d].location` refers to an invalid local file path: %q: %w", i, f.Location, err)
-			}
-			// f.Location does NOT need to be accessible, so we do NOT check os.Stat(f.Location)
+		if err := validateFile(f.File, fmt.Sprintf("images[%d]", i)); err != nil {
+			return err
 		}
-		switch f.Arch {
-		case X8664, AARCH64:
-		default:
-			return fmt.Errorf("field `images.arch` must be %q or %q, got %q", X8664, AARCH64, f.Arch)
+		if f.Kernel != nil {
+			if err := validateFile(f.Kernel.File, fmt.Sprintf("images[%d].kernel", i)); err != nil {
+				return err
+			}
```

Comment:
Should there be a requirement that the kernel arch matches the image arch?

Same question for Initrd.

Given that all these are loaded before binfmt_misc, I don't see how you could mix architectures between those 3 files.

Reply:
fixed